slotsmanager.c

```diff
@@ -0,0 +1,427 @@
+#include <string.h>
+#include "allvars.h"
+#include "slotsmanager.h"
+#include "mymalloc.h"
+#include "timestep.h"
+#include "system.h"
+#include "endrun.h"
+#include "openmpsort.h"
+#include "forcetree.h"
+
+struct slots_manager_type SlotsManager[1] = {0};
+
+MPI_Datatype MPI_TYPE_PARTICLE = 0;
+MPI_Datatype MPI_TYPE_PLAN_ENTRY = 0;
+MPI_Datatype MPI_TYPE_SLOT[6] = {0};
+
+static struct sph_particle_data * GDB_SphP;
+static struct star_particle_data * GDB_StarP;
+static struct bh_particle_data * GDB_BhP;
+
+static int
+slots_gc_base();
+
+static int
+slots_gc_slots();
+
+int
+slots_fork(int parent, int ptype)
+{
+    /* this will fork a zero mass particle at the given location of parent of the given type.
+     *
+     * Assumes the particle is protected by locks in threaded env.
+     *
+     * The Generation of parent is incremented.
+     * The child carries the incremented generation number.
+     * The ID of the child is modified, with the new generation number set
+     * at the highest 8 bits.
+     *
+     * the new particle's index is returned.
+     *
+     * Its mass and ptype can be then adjusted. (watchout detached BH /SPH
+     * data!)
+     * PI will point to a new slot for this type.
+     * if the slots runs out, this will trigger a slots growth
+     * */
+
+    if(NumPart >= All.MaxPart)
+    {
+        endrun(8888,
+                "On Task=%d with NumPart=%d we try to spawn. Sorry, no space left...(All.MaxPart=%d)\n",
+                ThisTask, NumPart, All.MaxPart);
+    }
+    /*This is all racy if ActiveParticle or P is accessed from another thread*/
+    int child = atomic_fetch_and_add(&NumPart, 1);
+    /*Update the active particle list:
+     * if the parent is active the child should also be active.
+     * Stars must always be active on formation, but
+     * BHs need not be: a halo can be seeded when the particle in question is inactive.*/
+    if(is_timebin_active(P[parent].TimeBin, All.Ti_Current)) {
+        int childactive = atomic_fetch_and_add(&NumActiveParticle, 1);
+        ActiveParticle[childactive] = child;
+    }
+
+    P[parent].Generation ++;
+    uint64_t g = P[parent].Generation;
+    /* change the child ID according to the generation. */
+    P[child] = P[parent];
+    P[child].ID = (P[parent].ID & 0x00ffffffffffffffL) + (g << 56L);
+
+    P[child].Mass = 0;
+    P[child].Type = ptype;
+
+    if(SlotsManager->info[ptype].enabled) {
+        /* if enabled, alloc a new Slot for secondary data */
+        int PI = atomic_fetch_and_add(&SlotsManager->info[ptype].size, 1);
+
+        if(PI >= SlotsManager->info[ptype].maxsize) {
+            /* rare case, use an expensive critical section */
+            #pragma omp critical
```

Comment:
Right, this will not work with the memory allocator. Also the critical section is not sufficient, because not all accesses to SlotsManager are guarded by a similar critical section. 

This case is very rare: until recently we did not even check for it and we have never seen a problem. I would drop the code.

Reply:
I though the critical is sufficient, for all access to info[ptype].maxsize is protected by it. The current code will endrun if a problem occurs -- so probably just drop the critical and grow?

Reply:
The critical isn't enough because we have the memmove in slots_reserve: this will be bad if we are accessing BhP at the same time. I don't know if there is any reasonable way around this without a stop_the_world() routine...

Reply:
That's correct. The access to the secondary data is not atomic. I'll change this to death spot on.